### PR TITLE
Hmap u64

### PR DESCRIPTION
**Pull request by ubitux**

### CODE REVIEW

File: libnopegl/src/hmap.c [hunk 1]
```diff
@@ -54,7 +64,27 @@ void ngli_hmap_set_free_func(struct hmap *hm, ngli_user_free_func_type user_free
 #define NO_REF (struct hmap_ref){.bucket_id = -1}
 #define HAS_REF(ref) ((ref).bucket_id != -1)
 
-struct hmap *ngli_hmap_create(void)
+static uint32_t key_hash_str(union hmap_key x) { return ngli_crc32(x.str); }
+static uint32_t key_hash_u64(union hmap_key x) { return ngli_crc32_mem(x.u8_8, sizeof(x.u8_8)); }
```
Comment (mbouron) [inline libnopegl/src/hmap.c:68]: Personal preference but ... can  you avoid interleaving the declaration ? it's hard to follow, especially if there is allocation involved.

**Review by mbouron — APPROVED** [review]
Except from my comment, LGTM.

File: libnopegl/src/hmap.c [hunk 2]
```diff
@@ -118,17 +118,17 @@ static struct hmap_ref ref_from_entry(const struct hmap *hm, const struct hmap_e
...
 }
 
 static struct hmap_entry *fixed_entry_from_ref(const struct hmap *hm, struct hmap_ref ref,
-                                               struct hmap_ref removed, int current_id)
+                                               struct hmap_ref removed, size_t current_id)
 {
     if (!HAS_REF(ref))
         return NULL;
-    const int need_fix =
+    const size_t need_fix =
```
Comment (mbouron) [inline libnopegl/src/hmap.c:131]: It looks like need_fix is a boolean. Isn't better to cast it later on ? or simplify use a `if` statement and sub explictely by `1`.

File: libnopegl/src/hmap.c (hunk 1 above)
Comment (ubitux) [reply libnopegl/src/hmap.c:68]: Batching functions with the same role is IMO way more readable and less error prone than the following:

```c
static uint32_t key_hash_str(union hmap_key x) { return ngli_crc32(x.str); }
static int key_cmp_str(union hmap_key a, union hmap_key b) { return strcmp(a.str, b.str); }
static union hmap_key key_dup_str(union hmap_key x) { return (union hmap_key){.str=ngli_strdup(x.str)}; }
static int key_check_str(union hmap_key x) { return !!x.str; }
static void key_free_str(union hmap_key x) { ngli_free(x.str); }

static uint32_t key_hash_u64(union hmap_key x) { return ngli_crc32_mem(x.u8_8, sizeof(x.u8_8)); }
static int key_cmp_u64(union hmap_key a, union hmap_key b) { return a.u64 != b.u64; }
static union hmap_key key_dup_u64(union hmap_key x) { return x; }
static int key_check_u64(union hmap_key x) { return 1; }
static void key_free_u64(union hmap_key x) { }
```

File: libnopegl/src/hmap.c (hunk 2 above)
Comment (ubitux) [reply libnopegl/src/hmap.c]: Yep, done with a cast

File: libnopegl/src/hmap.c (hunk 1 above)
Comment (mbouron) [reply libnopegl/src/hmap.c:68]: Again it's personal preference. Since you are the main contributor to this module, you can keep it as-is.

**Review by mbouron — APPROVED** [review]
